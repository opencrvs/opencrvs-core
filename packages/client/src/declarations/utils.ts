/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { EVENT_STATUS } from '@client/workqueue'
import { SUBMISSION_STATUS } from '.'

export const isPendingCorrection = (status: SUBMISSION_STATUS | undefined) =>
  status === EVENT_STATUS.CORRECTION_REQUESTED

export const isReviewableDeclaration = (
  status: SUBMISSION_STATUS | undefined
) => status && [EVENT_STATUS.DECLARED, EVENT_STATUS.VALIDATED].includes(status)

export const isUpdatableDeclaration = (status: SUBMISSION_STATUS | undefined) =>
  status &&
  [
    SUBMISSION_STATUS.DRAFT,
    EVENT_STATUS.IN_PROGRESS,
    EVENT_STATUS.REJECTED
  ].includes(status)

export const isPrintable = (status: SUBMISSION_STATUS | undefined) =>
  status &&
  [SUBMISSION_STATUS.REGISTERED, SUBMISSION_STATUS.ISSUED].includes(status)

export const isIssuable = (status: SUBMISSION_STATUS | undefined) =>
  status === SUBMISSION_STATUS.CERTIFIED

export const isRecordOrDeclaration = (status: SUBMISSION_STATUS | undefined) =>
  status
    ? [
        SUBMISSION_STATUS.REGISTERED,
        SUBMISSION_STATUS.CORRECTION_REQUESTED,
        SUBMISSION_STATUS.CERTIFIED
      ].includes(status)
      ? 'record'
      : 'declaration'
    : ''

export const canBeCorrected = (status: SUBMISSION_STATUS | undefined) =>
  status &&
  [
    SUBMISSION_STATUS.REGISTERED,
    SUBMISSION_STATUS.CERTIFIED,
    SUBMISSION_STATUS.ISSUED
  ].includes(status)

export const isArchivable = (status: SUBMISSION_STATUS | undefined) =>
  status &&
  [
    SUBMISSION_STATUS.IN_PROGRESS,
    SUBMISSION_STATUS.DECLARED,
    SUBMISSION_STATUS.VALIDATED,
    SUBMISSION_STATUS.REJECTED
  ].includes(status)

export const canBeReinstated = (status: SUBMISSION_STATUS | undefined) =>
  status === SUBMISSION_STATUS.ARCHIVED

export const isViewable = (status: SUBMISSION_STATUS | undefined) =>
  status !== SUBMISSION_STATUS.DRAFT
