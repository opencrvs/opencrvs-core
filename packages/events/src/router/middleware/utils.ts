/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import type { AnyTRPCMiddlewareFunction } from '@trpc/server'
import {
  ActionInputWithType,
  TokenWithBearer,
  TokenUserType
} from '@opencrvs/commons'

export interface UserDetails {
  type: TokenUserType.USER | TokenUserType.SYSTEM
  id: string
  primaryOfficeId: string | undefined
  role: string
}

export type Context = { token: TokenWithBearer; user: UserDetails }
/**
 * TRPC Middleware options with correct context.
 * Actual middleware type definition is only for internal use within TRPC.
 */
export type MiddlewareOptions = Omit<
  Parameters<AnyTRPCMiddlewareFunction>[0],
  'ctx'
> & { ctx: Context }

export type ActionMiddlewareOptions = Omit<MiddlewareOptions, 'input'> & {
  input: ActionInputWithType
}
