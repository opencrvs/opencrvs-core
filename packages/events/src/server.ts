/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { createServer, IncomingMessage } from 'http'
import { createOpenApiHttpHandler } from 'trpc-to-openapi'
import { createHTTPHandler } from '@trpc/server/adapters/standalone'
import '@opencrvs/commons/monitoring'
import { logger } from '@opencrvs/commons'
import { appRouter } from './router/router'
import { createContext } from './context'

function stringifyRequest(req: IncomingMessage) {
  const url = new URL(req.url || '', `http://${req.headers.host}`)
  return `'${req.method} ${url.pathname}'`
}

const trpcConfig: Parameters<typeof createHTTPHandler>[0] = {
  router: appRouter,
  middleware: (req, _, next) => {
    logger.info(`Request: ${stringifyRequest(req)}`)
    return next()
  },
  onError: ({ error, req }) => {
    logger.warn(
      `Error for request: ${stringifyRequest(req)}. Error: '${error.message}'`
    )
  },
  createContext
}

function isTrpcRequest(req: IncomingMessage): boolean {
  if (!req.url) {
    throw new Error('No URL provided')
  }
  const url = new URL(req.url, `http://${req.headers.host}`)
  const pathName = url.pathname.replace(/^\//, '')
  const trpcProcedurePaths = Object.keys(appRouter._def.procedures)

  return (
    url.search.startsWith('?input') || trpcProcedurePaths.includes(pathName)
  )
}

export function server() {
  const restServer = createOpenApiHttpHandler(trpcConfig)
  const trpcServer = createHTTPHandler(trpcConfig)

  return createServer((req, res) => {
    if (!req.url) {
      res.writeHead(500)
      res.end('No URL provided')
      return
    }

    // If it's a tRPC request, handle it with the tRPC server
    if (isTrpcRequest(req)) {
      trpcServer(req, res)
    } else {
      // If it's a REST request, handle it with the REST server
      void restServer(req, res)
    }
  })
}
