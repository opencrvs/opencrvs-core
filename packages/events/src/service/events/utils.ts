/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { Action, ActionType } from '@opencrvs/commons'

export function findLastAssignmentAction(actions: Action[]) {
  return actions
    .filter(
      ({ type }) => type === ActionType.ASSIGN || type === ActionType.UNASSIGN
    )
    .reduce<
      Action | undefined
    >((latestAction, action) => (!latestAction || action.createdAt > latestAction.createdAt ? action : latestAction), undefined)
}
