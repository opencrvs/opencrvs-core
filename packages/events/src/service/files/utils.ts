/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import { differenceWith, first, isEqual, orderBy } from 'lodash'
import {
  Draft,
  EventConfig,
  EventDocument,
  FieldConfig,
  FileFieldType,
  getActionAnnotationFields,
  AnyFileField,
  getDeclarationFields,
  getMixedPath,
  FileFieldWithOptionValue,
  FileFieldValue,
  ActionUpdate,
  getCurrentEventStateWithDrafts,
  EventState,
  getCurrentEventState,
  ActionDocument
} from '@opencrvs/commons'

function isFieldWithFile(field: FieldConfig): field is AnyFileField {
  return FileFieldType.safeParse(field.type).success
}

/**
 *
 * @param field File field
 * @param update single action or event state.
 * @returns normalised file reference based on the field id and event state.
 */
function getFileReferences(
  field: AnyFileField,
  update: ActionUpdate | EventState
): { id: string; value: string }[] {
  const value = getMixedPath<AnyFileField['defaultValue']>(update, field.id)

  if (!value) {
    return []
  }

  const maybeFileOptions = FileFieldWithOptionValue.safeParse(value)
  if (maybeFileOptions.success) {
    return maybeFileOptions.data.map((opt) => ({
      id: field.id,
      value: opt.path
    }))
  }

  const maybeFile = FileFieldValue.safeParse(value)

  if (maybeFile.success) {
    return [{ id: field.id, value: maybeFile.data.path }]
  }

  if (typeof value === 'string') {
    return [{ id: field.id, value }]
  }

  throw new Error(
    `Unknown filetype, ${field.type} with value: ${JSON.stringify(value)}`
  )
}

/**
 *  Since API allows sending PATCH requests with only the changed fields,
 *  we need to check the current event state with the current draft.
 * @returns file references that are not present in the current event state but are when previous draft is applied to the state.
 */
function getUnreferencedDeclarationDraftFiles({
  event,
  currentDraft,
  previousDraft,
  configuration
}: {
  event: EventDocument
  currentDraft?: Draft
  previousDraft: Draft
  configuration: EventConfig
}) {
  const currentEventState = currentDraft
    ? getCurrentEventStateWithDrafts({
        event,
        draft: currentDraft,
        configuration
      })
    : getCurrentEventState(event, configuration)

  const previousEventState = getCurrentEventStateWithDrafts({
    event,

    draft: {
      // Given previous draft, if event has been updated afterwards (running deletion after submission of actual action.) the data would never get applied.
      ...previousDraft,
      createdAt: new Date().toISOString(),
      action: {
        ...previousDraft.action,
        createdAt: new Date().toISOString()
      }
    },
    configuration
  })

  const declarationFileFields =
    getDeclarationFields(configuration).filter(isFieldWithFile)

  const currentDeclarationFiles = declarationFileFields.flatMap((field) =>
    getFileReferences(field, currentEventState.declaration)
  )

  const previousDeclarationFiles = declarationFileFields.flatMap((field) =>
    getFileReferences(field, previousEventState.declaration)
  )

  return differenceWith(
    previousDeclarationFiles,
    currentDeclarationFiles,
    isEqual
  )
}

function getUnreferencedAnnotationDraftFiles({
  currentAction,
  previousDraft,
  configuration
}: {
  currentAction: Draft['action'] | ActionDocument
  previousDraft: Draft
  configuration: EventConfig
}) {
  if (currentAction.type !== previousDraft.action.type) {
    // eslint-disable-next-line no-console
    console.warn(
      'Previous draft was for different action. Not comparing file references between them.'
    )
    return []
  }

  const currentDraftActionConfig = configuration.actions.find(
    (action) => action.type === previousDraft.action.type
  )

  if (!currentDraftActionConfig) {
    // eslint-disable-next-line no-console
    console.warn(
      `No configuration found for action ${previousDraft.action.type}.`
    )
  }

  const annotationFileFields = currentDraftActionConfig
    ? getActionAnnotationFields(currentDraftActionConfig).filter(
        isFieldWithFile
      )
    : []

  // @TODO: There might be a case where annotation shares id with declaration field? Should it limited by the configuration?
  const currentAnnotationFiles = annotationFileFields.flatMap((field) =>
    getFileReferences(field, currentAction.annotation ?? {})
  )

  const previousAnnotationFiles = annotationFileFields.flatMap((field) =>
    getFileReferences(field, previousDraft.action.annotation ?? {})
  )

  return differenceWith(
    previousAnnotationFiles,
    currentAnnotationFiles,
    isEqual
  )
}

/**
 *
 * @returns all the filenames that are referenced in the current draft, but not in the previous draft.
 */
export function getUnreferencedDraftFiles({
  event,
  currentDraft,
  previousDraft,
  configuration
}: {
  event: EventDocument
  /**
   * Current draft is optional, because we might be comparing the previous draft with the current event state. (e.g. actual action has been submitted)
   */
  currentDraft?: Draft
  previousDraft: Draft
  configuration: EventConfig
}) {
  const unreferencedDeclarationDraftFiles =
    getUnreferencedDeclarationDraftFiles({
      event,
      currentDraft,
      previousDraft,
      configuration
    })

  const currentAction =
    currentDraft?.action ??
    first(
      orderBy(
        event.actions.filter(
          (action) => action.type === previousDraft.action.type
        ),
        ['createdAt'],
        'desc'
      )
    )

  if (!currentAction) {
    // This should not happen. It means that the previous draft action type was invalid.
    // eslint-disable-next-line no-console
    console.error(
      `Could not find corresponding action ${previousDraft.action.type} for event: ${event.id}`
    )

    return unreferencedDeclarationDraftFiles
  }

  const unreferencedAnnotationDraftFiles = getUnreferencedAnnotationDraftFiles({
    // @TODO: for some reason declaration is omitted from some type?
    currentAction: currentAction as Draft['action'] | ActionDocument,
    previousDraft,
    configuration
  })

  return [
    ...unreferencedDeclarationDraftFiles,
    ...unreferencedAnnotationDraftFiles
  ]
}
