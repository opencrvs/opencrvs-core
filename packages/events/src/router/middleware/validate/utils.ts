/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { TRPCError } from '@trpc/server'
import _ from 'lodash'
import {
  ActionUpdate,
  FieldConfig,
  getFieldValidationErrors,
  Inferred,
  errorMessages
} from '@opencrvs/commons/events'

type ValidationError = {
  message: string
  id: string
  value: unknown
}

export function getFormFieldErrors(formFields: Inferred[], data: ActionUpdate) {
  return formFields.reduce(
    (errorResults: ValidationError[], field: FieldConfig) => {
      const fieldErrors = getFieldValidationErrors({
        field,
        values: data
      }).errors

      if (fieldErrors.length === 0) {
        return errorResults
      }

      // For backend, use the default message without translations.
      const errormessageWithId = fieldErrors.map((error) => ({
        message: error.message.defaultMessage,
        id: field.id,
        value: data[field.id as keyof typeof data]
      }))

      return [...errorResults, ...errormessageWithId]
    },
    []
  )
}

export function getVerificationPageErrors(
  verificationPageIds: string[],
  data: ActionUpdate
) {
  return verificationPageIds
    .map((pageId) => {
      const value = data[pageId]
      return typeof value !== 'boolean'
        ? {
            message: 'Verification page result is required',
            id: pageId,
            value
          }
        : null
    })
    .filter((error) => error !== null)
}

export function throwWhenNotEmpty(errors: unknown[]) {
  if (errors.length > 0) {
    throw new TRPCError({
      code: 'BAD_REQUEST',
      message: JSON.stringify(errors)
    })
  }
}

function isObject(value: unknown): value is Record<string, unknown> {
  return _.isPlainObject(value)
}

/**
 * Flattens an object into a list of [key, value] pairs using dot/bracket notation.
 */
function flattenEntries<T>(obj: T, path: string = ''): [string, unknown][] {
  if (Array.isArray(obj)) {
    return obj.flatMap((value, index) =>
      flattenEntries(value, `${path}[${index}]`)
    )
  }

  if (isObject(obj)) {
    return Object.entries(obj).flatMap(([key, value]) =>
      flattenEntries(value, path ? `${path}.${key}` : key)
    )
  }

  return obj === undefined ? [] : [[path, obj]]
}

/**
 * Compares update vs cleaned payloads and returns an array of offending keys.
 */
export function getInvalidUpdateKeys<T>({
  update,
  cleaned
}: {
  update: T
  cleaned: T
}): ValidationError[] {
  const updateEntries = flattenEntries(update)
  const cleanedKeys = flattenEntries(cleaned).map(([key]) => key)

  return updateEntries
    .filter(([key]) => !cleanedKeys.includes(key))
    .map(([key, value]) => ({
      message: errorMessages.hiddenField.defaultMessage,
      id: key,
      value
    }))
}
