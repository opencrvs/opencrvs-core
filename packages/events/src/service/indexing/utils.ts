/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import { EventSearchIndex } from '@opencrvs/commons/events'
import {
  QueryDslQueryContainer,
  SearchRequest
} from '@elastic/elasticsearch/lib/api/types'

const FIELD_SEPARATOR = '____'
export const DEFAULT_SIZE = 10

/**
 * Generates an Elasticsearch query to search within `document.data`
 * using the provided search payload.
 */
export function generateQuery(event: Omit<EventSearchIndex, 'type'>) {
  const must: QueryDslQueryContainer[] = Object.entries(event).map(
    ([key, value]) => ({
      match: {
        [`data.${key.replaceAll('.', FIELD_SEPARATOR)}`]: value
      }
    })
  )

  return {
    bool: {
      must
    }
  } as SearchRequest['query']
}
