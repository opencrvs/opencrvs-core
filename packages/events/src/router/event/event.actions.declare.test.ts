/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { TRPCError } from '@trpc/server'
import {
  ActionType,
  AddressType,
  generateActionDeclarationInput,
  getAcceptedActions,
  SCOPES
} from '@opencrvs/commons'
import { tennisClubMembershipEvent } from '@opencrvs/commons/fixtures'
import { createTestClient, setupTestCase } from '@events/tests/utils'

test(`prevents forbidden access if missing required scope`, async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user, [])

  await expect(
    client.event.actions.declare.request(
      generator.event.actions.declare('event-test-id-12345', {})
    )
  ).rejects.toMatchObject(new TRPCError({ code: 'FORBIDDEN' }))
})

test(`allows access if required scope is present`, async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user, [SCOPES.RECORD_DECLARE])

  await expect(
    client.event.actions.declare.request(
      generator.event.actions.declare('event-test-id-12345', {})
    )
  ).rejects.not.toMatchObject(new TRPCError({ code: 'FORBIDDEN' }))
})

test('Validation error message contains all the offending fields', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const event = await client.event.create(generator.event.create())

  const data = generator.event.actions.declare(event.id, {
    declaration: {
      'applicant.dob': '02-02',
      'applicant.dobUnknown': false,
      'recommender.none': true
    }
  })

  await expect(
    client.event.actions.declare.request(data)
  ).rejects.matchSnapshot()
})

test('when mandatory field is invalid, conditional hidden fields are still skipped', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const event = await client.event.create(generator.event.create())

  const data = generator.event.actions.declare(event.id, {
    declaration: {
      'applicant.dob': '02-1-2024',
      'applicant.dobUnknown': false,
      'applicant.firstname': 'John',
      'applicant.surname': 'Doe',
      'recommender.none': true,
      'applicant.address': {
        country: 'FAR',
        addressType: AddressType.DOMESTIC,
        province: 'a45b982a-5c7b-4bd9-8fd8-a42d0994054c',
        district: '5ef450bc-712d-48ad-93f3-8da0fa453baa',
        urbanOrRural: 'RURAL' as const,
        village: 'Small village'
      }
    }
  })

  await expect(
    client.event.actions.declare.request(data)
  ).rejects.matchSnapshot()
})

test('Skips required field validation when they are conditionally hidden', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const event = await client.event.create(generator.event.create())

  const form = {
    'applicant.dob': '2024-02-01',
    'applicant.dobUnknown': false,
    'applicant.firstname': 'John',
    'applicant.surname': 'Doe',
    'recommender.none': true,
    'applicant.address': {
      country: 'FAR',
      addressType: AddressType.DOMESTIC,
      province: 'a45b982a-5c7b-4bd9-8fd8-a42d0994054c',
      district: '5ef450bc-712d-48ad-93f3-8da0fa453baa',
      urbanOrRural: 'RURAL' as const,
      village: 'Small village'
    }
  }

  const data = generator.event.actions.declare(event.id, {
    declaration: form
  })

  const response = await client.event.actions.declare.request(data)
  const activeActions = getAcceptedActions(response)

  const savedAction = activeActions.find(
    (action) => action.type === ActionType.DECLARE
  )
  expect(savedAction?.declaration).toEqual(form)
})

test('gives validation error when a conditional page, which is visible, has a required field', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const event = await client.event.create(generator.event.create())

  const form = {
    // When the applicant.dob is before 1950-01-01, the senior-pass.id field on senior-pass page is required
    'applicant.dob': '1944-02-01',
    'applicant.firstname': 'John',
    'applicant.surname': 'Doe',
    'recommender.none': true,
    'applicant.address': {
      country: 'FAR',
      addressType: AddressType.DOMESTIC,
      province: 'a45b982a-5c7b-4bd9-8fd8-a42d0994054c',
      district: '5ef450bc-712d-48ad-93f3-8da0fa453baa',
      urbanOrRural: 'RURAL' as const,
      village: 'Small village'
    }
  }

  const data = generator.event.actions.declare(event.id, {
    declaration: form
  })

  await expect(
    client.event.actions.declare.request(data)
  ).rejects.matchSnapshot()
})

test('successfully validates a fields on a conditional page, which is visible', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const event = await client.event.create(generator.event.create())

  const form = {
    // When the applicant.dob is before 1950-01-01, the senior-pass.id field on senior-pass page is required
    'applicant.dob': '1944-02-01',
    'senior-pass.id': '1234567890',
    'applicant.firstname': 'John',
    'applicant.surname': 'Doe',
    'recommender.none': true,
    'applicant.address': {
      country: 'FAR',
      addressType: AddressType.DOMESTIC,
      province: 'a45b982a-5c7b-4bd9-8fd8-a42d0994054c',
      district: '5ef450bc-712d-48ad-93f3-8da0fa453baa',
      urbanOrRural: 'RURAL' as const,
      village: 'Small village'
    }
  }

  const data = generator.event.actions.declare(event.id, {
    declaration: form
  })

  const response = await client.event.actions.declare.request(data)
  const activeActions = getAcceptedActions(response)

  const savedAction = activeActions.find(
    (action) => action.type === ActionType.DECLARE
  )
  expect(savedAction?.declaration).toEqual(form)
})

test('Prevents adding birth date in future', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const event = await client.event.create(generator.event.create())

  const form = {
    'applicant.dob': '2040-02-01',
    'applicant.dobUnknown': false,
    'applicant.firstname': 'John',
    'applicant.surname': 'Doe',
    'recommender.none': true,
    'applicant.address': {
      country: 'FAR',
      addressType: AddressType.DOMESTIC,
      province: 'a45b982a-5c7b-4bd9-8fd8-a42d0994054c',
      district: '5ef450bc-712d-48ad-93f3-8da0fa453baa',
      urbanOrRural: 'RURAL' as const,
      village: 'Small village'
    }
  }

  const payload = generator.event.actions.declare(event.id, {
    declaration: form
  })

  await expect(
    client.event.actions.declare.request(payload)
  ).rejects.matchSnapshot()
})

test('validation prevents including hidden fields', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const event = await client.event.create(generator.event.create())

  const data = generator.event.actions.declare(event.id, {
    declaration: {
      ...generateActionDeclarationInput(
        tennisClubMembershipEvent,
        ActionType.DECLARE
      ),
      'recommender.firstname': 'this should not be here'
    }
  })

  await expect(
    client.event.actions.declare.request(data)
  ).rejects.matchSnapshot()
})

test('valid action is appended to event actions', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const event = await client.event.create(generator.event.create())

  const data = generator.event.actions.declare(event.id)

  await client.event.actions.declare.request(data)
  const updatedEvent = await client.event.get(event.id)

  expect(updatedEvent.actions).toEqual([
    expect.objectContaining({ type: ActionType.CREATE }),
    expect.objectContaining({ type: ActionType.ASSIGN }),
    expect.objectContaining({
      type: ActionType.DECLARE
    }),
    expect.objectContaining({ type: ActionType.UNASSIGN }),
    expect.objectContaining({
      type: ActionType.READ
    })
  ])
})
