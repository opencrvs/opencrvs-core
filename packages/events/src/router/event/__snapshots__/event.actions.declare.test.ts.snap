// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Prevents adding birth date in future 1`] = `[TRPCError: [{"message":"Please enter a valid date","id":"applicant.dob","value":"2040-02-01"}]]`;

exports[`Validation error message contains all the offending fields 1`] = `[TRPCError: [{"message":"Required","id":"applicant.name"},{"message":"Invalid date field","id":"applicant.dob","value":"02-02"},{"message":"Please enter a valid date","id":"applicant.dob","value":"02-02"},{"message":"Required","id":"applicant.address"}]]`;

exports[`gives validation error when a conditional page, which is visible, has a required field 1`] = `[TRPCError: [{"message":"Required","id":"senior-pass.id"}]]`;

exports[`validation prevents including hidden fields 1`] = `[TRPCError: [{"message":"Hidden or disabled field should not receive a value","id":"recommender.firstname","value":"this should not be here"}]]`;

exports[`when mandatory field is invalid, conditional hidden fields are still skipped 1`] = `[TRPCError: [{"message":"Invalid date field","id":"applicant.dob","value":"02-1-2024"},{"message":"Please enter a valid date","id":"applicant.dob","value":"02-1-2024"}]]`;
